crates/pathfinder/src/rpc/v02/method/get_transaction_by_block_id_and_index.rs

```diff
@@ -0,0 +1,256 @@
+use crate::rpc::v02::RpcContext;
+use crate::storage::{StarknetBlocksBlockId, StarknetBlocksTable, StarknetTransactionsTable};
+
+#[derive(serde::Deserialize, Debug, PartialEq, Eq)]
+pub struct GetTransactionByBlockIdAndIndexInput {
+    block_id: BlockId,
+    index: StarknetTransactionIndex,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(
+    GetTransactionByBlockIdAndIndexError: BlockNotFound,
+    InvalidTxnIndex
+);
+
+#[allow(dead_code)]
+pub async fn get_transaction_by_block_id_and_index(
+    context: std::sync::Arc<RpcContext>,
+    input: GetTransactionByBlockIdAndIndexInput,
+) -> Result<Transaction, GetTransactionByBlockIdAndIndexError> {
+    let index: usize = input
+        .index
+        .get()
+        .try_into()
+        .map_err(|_| GetTransactionByBlockIdAndIndexError::InvalidTxnIndex)?;
+
+    let block_id = match input.block_id {
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+        BlockId::Latest => StarknetBlocksBlockId::Latest,
+        BlockId::Pending => {
+            return get_transaction_from_pending(&context.pending_data, index).await
+        }
+    };
+
+    let storage = context.storage.clone();
+    let span = tracing::Span::current();
+
+    let jh = tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut db = storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let db_tx = db.transaction().context("Creating database transaction")?;
+
+        // Get the transaction from storage.
+        match StarknetTransactionsTable::get_transaction_at_block(&db_tx, block_id, index)
+            .context("Reading transaction from database")?
+        {
+            Some(transaction) => Ok(transaction.into()),
+            None => {
+                // We now need to check whether it was the block hash or transaction index which were invalid. We do this by checking if the block exists
+                // at all. If no, then the block hash is invalid. If yes, then the index is invalid.
+                //
+                // get_root is cheaper than querying the full block.
+                match StarknetBlocksTable::get_root(&db_tx, block_id)
+                    .context("Reading block from database")?
+                {
+                    Some(_) => Err(GetTransactionByBlockIdAndIndexError::InvalidTxnIndex),
+                    None => Err(GetTransactionByBlockIdAndIndexError::BlockNotFound),
+                }
+            }
+        }
+    });
+
+    jh.await.context("Database read panic or shutting down")?
+}
+
+async fn get_transaction_from_pending(
+    pending: &Option<crate::state::PendingData>,
+    index: usize,
+) -> Result<Transaction, GetTransactionByBlockIdAndIndexError> {
+    match pending {
+        Some(pending) => pending.block().await.map_or_else(
+            || Err(GetTransactionByBlockIdAndIndexError::BlockNotFound),
```

Comment:
We should check what the expected behaviour for this is. This would also impact the `None` arm. Other options:

- default to `latest` if no pending is available (doesn't really make sense though)
- return `InvalidTxIndex` instead - the intent being that we have 0 pending transactions available

Rational being that from a user perspective there is always a "pending" block.

Reply:
That's a good point.
Defaulting to latest doesn't make sense I think. Maybe we should clarify this behavior with Starkware?

Reply:
Based on discussion with Starkware I'll go forward with invalid transaction index.